File: Python/other/Maximum_Square_Matrix.py
# find the size of Maximum Square Matrix With All 1s


def MaxSquare(mat):
    # creating a null matrix
	ans = 0
    temp = [[0] * len(mat[0])] * len(mat)
    # Traversing through the given matrix
    for i in range(len(mat)):
        for j in range(len(mat[0])):
            temp[i][j] = mat[i][j]
            # checking if it is not a corner case:
            if i > 0 and j > 0 and mat[i][j] == 1:
                temp[i][j] = min(temp[i][j - 1], temp[i - 1][j], temp[i - 1][j - 1])
				temp[i][j] += 1
            # for finding max size of square mat
            if ans < temp[i][j]:
                ans = temp[i][j]
    return ans


mat = []
rows = input("Enter rows")
col = input("Enter col")
for i in range(0, int(rows)):
    mat.append([int(j) for j in input().split()])

print("output is", MaxSquare(mat))
'''
output:
Enter rows3
Enter col3
0 0 0
0 1 0
0 0 0

output is 1

Enter rows3
Enter col3
0 1 1
0 1 1
0 0 0

output is 2

Time Complexity : O(rows*col)'''

File: Java/other/MaximumSquareMatrix.java
import java.util.Scanner;

public class MaximumSquareMatrix {

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter number of rows and columns:");
        int r = scanner.nextInt();
        int c = scanner.nextInt();
        int[][] matrix = new int[r][c];
        System.out.println("Enter elements of the matrix:");
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                matrix[i][j] = scanner.nextInt();
            }
        }

        printMaxSquare(matrix);
    }

    private static void printMaxSquare(int[][] matrix) {

        int rows = matrix.length, cols = matrix[0].length, i, j, maxSquare, max_i, max_j;
        int[][] squareSizes = new int[rows][cols];

        // Set first column of squareSizes[][]
        for(i = 0; i < rows; i++) {
            squareSizes[i][0] = matrix[i][0];
        }

        // Set first row of squareSizes[][]
        for(j = 0; j < cols; j++) {
            squareSizes[0][j] = matrix[0][j];
        }

        // getting other entries for squareSizes
        for (i = 1; i < rows; i++) {
            for (j = 1; j < cols; j++) {
                if(matrix[i][j] == 1) {
                    squareSizes[i][j] = Math.min(squareSizes[i][j - 1],
                            Math.min(squareSizes[i - 1][j], squareSizes[i - 1][j - 1])) + 1;
                } else {
                    squareSizes[i][j] = 0;
                }
            }
        }

        // find max entry of squareSizes
        maxSquare = squareSizes[0][0]; max_i = 0; max_j = 0;
        for(i = 0; i < rows; i++)
        {
            for(j = 0; j < cols; j++)
            {
                if(maxSquare < squareSizes[i][j])
                {
                    maxSquare = squareSizes[i][j];
                    max_i = i;
                    max_j = j;
                }
            }
        }

        // print result
        System.out.println("Maximum size sub-matrix is: ");
        for(i = max_i; i > max_i - maxSquare; i--)
        {
            for(j = max_j; j > max_j - maxSquare; j--)
            {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
    }
}
/*
* Sample input/output:
* Enter number of rows and columns:
* 6 5
* Enter elements of the matrix:
* 0 1 1 0 1
* 1 1 0 1 0
* 0 1 1 1 0
* 1 1 1 1 0
* 1 1 1 1 1
* 0 0 0 0 0
* Maximum size sub-matrix is:
* 1 1 1
* 1 1 1
* 1 1 1
*
* Space and Time complexity = O(m*n)
*/
